protocols/noise/src/lib.rs

```diff
@@ -319,23 +320,20 @@ where
     NoiseConfig<P, C, R>: UpgradeInfo + InboundUpgrade<T,
         Output = (RemoteIdentity<C>, NoiseOutput<Negotiated<T>>),
         Error = NoiseError
-    >,
+    > + 'static,
     T: AsyncRead + AsyncWrite + Send + 'static,
     C: Protocol<C> + AsRef<[u8]> + Zeroize + Send + 'static,
+
 {
     type Output = (PeerId, NoiseOutput<Negotiated<T>>);
     type Error = NoiseError;
-    type Future = future::AndThen<
-        <NoiseConfig<P, C, R> as InboundUpgrade<T>>::Future,
-        FutureResult<Self::Output, Self::Error>,
-        fn((RemoteIdentity<C>, NoiseOutput<Negotiated<T>>)) -> FutureResult<Self::Output, Self::Error>
-    >;
+    type Future = Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Unpin>;
```

Comment:
```suggestion
    type Future = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
```

Reply:
@tomaka that would require `core::upgrade::InboundUpgrade::Future` to be `Send`. Should that be part of this pull request?

Reply:
We definitely want all futures to be send, as they can move around between threads of a threads pool while they are being executed, so yes.

Reply:
Sounds good. See https://github.com/libp2p/rust-libp2p/pull/1248/commits/555c2df315e44f21ad39d4408445ce2cb84dd1a4.